### PR TITLE
drivers: can: Wait for MCP2515 to reset

### PR DESCRIPTION
**Pull request by NZSmartie**
The device may be prematurely configured before it is ready during reset
without first checking. After the driver writes to the registers and
normal mode gets selected, CANSTAT is read back to ensure the device is
now configured in normal mode (0x00). However, an unresponsive device
will be read as 0x00, the driver then assumes it is configured correctly.

The MCP2515 will reset into configuration mode (0x04) to allow bit timings
to be configured. Adding a check for this mode prevents the device from
being misconfigured.

Signed-off-by: Roman Vaughan <nzsmartie@gmail.com>

### CODE REVIEW

File: drivers/can/can_mcp2515.c [hunk 1]
```diff
@@ -352,6 +364,21 @@ static int mcp2515_configure(struct device *dev, enum can_mode mode,
...
+		if (ret == MCP2515_MODE_CONFIGURATION) {
+			break;
+		}
+		if (ret < 0) {
+			LOG_ERR("Device did not reset into configuration mode [%d]",
+				ret);
+			return -EIO;
+		}
+
+		k_sleep(Z_TIMEOUT_TICKS(1));
```
Comment (alexanderwachter) [inline drivers/can/can_mcp2515.c:379]: if you want to sleep for one ms here, you need to use `k_msleep(1);`.

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

File: drivers/can/can_mcp2515.c [hunk 2]
```diff
@@ -271,6 +271,18 @@ const int mcp2515_set_mode(struct device *dev, uint8_t mcp2515_mode)
 	return 0;
 }
 
+static int mcp2515_get_mode(struct device *dev)
+{
+	u8_t canstat;
```
Comment (nixward) [inline drivers/can/can_mcp2515.c:276]: This should be uint8_t

Comment (NZSmartie) [reply drivers/can/can_mcp2515.c:276]: oh yeah, old habits!

**Review by nixward — CHANGES_REQUESTED** [review]
@NZSmartie can you please provide a reference in the MCP2515 datasheet where it says a delay is required after a soft reset before other SPI operations?  I'm not sure there is a problem.  The reset should be practically instant.

In the case that the MCP2515 chip is faulty or disconnected the added `while (1)` loop is not great as it will take down the main thread that calls `mcp2515_init()`.  The MC2515 can fail but it should not take down the calling thread.

**Comment by alexanderwachter** [comment]
@NZSmartie @nixward indeed. This should definitely timeout somewhen.

**Comment by NZSmartie** [comment]
> ## **8.1** Oscillator Start-up Timer
> The  MCP2515  utilizes  an  Oscillator  Start-up  Timer(OST) that holds the MCP2515 in Reset to ensure that the  oscillator  has  stabilized  before  the  internal  state machine begins to operate. The OST keeps the device in  a  Reset  state  for  128  OSC1  clock  cycles  after  the occurrence of a Power-on Reset, SPI Reset, after the assertion of the RESET pin, and after a wake-up from Sleep  mode.  It  should  be  noted  that  no  SPI  protocol operations should be attempted until after the OST has expired.

Instead of waiting for OST to expire, a query of the CANSTAT register is done to verify the device is in configuration mode (i.e. the device starts accepting SPI protocol operations). 128 OSC1 clock cycles may not seem long, but it does appear to have an affect on the system I'm developing for. 

As for the `while(1)` loop, i'll add a timeout mechanism

**Comment by nixward** [comment]
For brevity, something like this should be ok, @alexanderwachter @karstenkoenig?  128 clock cycles at 1MHz (minimum clock in frequency) is 128us.
```
/* 
 * Wait 128 OSC1 clock cycles at 1MHz (minimum clock in frequency)
 * see datasheet section 8.1 Oscillator Start-up Timer
 */
k_usleep(128); 

ret = mcp2515_get_mode(dev);
if (ret != MCP2515_MODE_CONFIGURATION) {
	LOG_ERR("Device did not reset into configuration mode [%d]",
		ret);
	return -EIO;
}
```

**Comment by NZSmartie** [comment]
> 
> 
> @NZSmartie @nixward indeed. This should definitely timeout somewhen.
> Maybe the power supply is not ready when the mcu initializes?

Possible, but I've measured all the rails come up and stabilise in roughly 600us. the MCU's power supply does come up before the MCP2515.

**Comment by NZSmartie** [comment]
> 
> 
> For brevity, something like this should be ok, @alexanderwachter @karstenkoenig? 128 clock cycles at 1MHz (minimum clock in frequency) is 128us.
> 
> ```
> /* 
>  * Wait 128 OSC1 clock cycles at 1MHz (minimum clock in frequency)
>  * see datasheet section 8.1 Oscillator Start-up Timer
>  */
> k_usleep(128); 
> 
> ret = mcp2515_get_mode(dev);
> if (ret != MCP2515_MODE_CONFIGURATION) {
> 	LOG_ERR("Device did not reset into configuration mode [%d]",
> 		ret);
> 	return -EIO;
> }
> ```

I would like to introduce more time for it to become available and not become a one shot race condition (what happens when the CAN controller's power comes up much later than the MCU?). Requiring the MCU to reset just to get talking with the CAN controller is not a valid fix.

**Comment by alexanderwachter** [comment]
@NZSmartie you can replace the while(1) with a for loop and abort after N cycles where N = 100ms / loop_delay or something like that.

File: drivers/can/can_mcp2515.c [hunk 3]
```diff
@@ -352,6 +364,29 @@ static int mcp2515_configure(struct device *dev, enum can_mode mode,
 		goto done;
 	}
 
+	/* Wait for MCP2515 to become ready. */
+	uint64_t now = k_uptime_get();
```
Comment (alexanderwachter) [inline drivers/can/can_mcp2515.c:368]: Variables need to be introduced at the beginning of a function.

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

**Comment by nixward** [comment]
> > 
> > 
> > For brevity, something like this should be ok, @alexanderwachter @karstenkoenig? 128 clock cycles at 1MHz (minimum clock in frequency) is 128us.
> > 
> > ```
> > /* 
> >  * Wait 128 OSC1 clock cycles at 1MHz (minimum clock in frequency)
> >  * see datasheet section 8.1 Oscillator Start-up Timer
> >  */
> > k_usleep(128); 
> > 
> > ret = mcp2515_get_mode(dev);
> > if (ret != MCP2515_MODE_CONFIGURATION) {
> > 	LOG_ERR("Device did not reset into configuration mode [%d]",
> > 		ret);
> > 	return -EIO;
> > }
> > ```
> 
> I would like to introduce more time for it to become available and not become a one shot race condition (what happens when the CAN controller's power comes up much later than the MCU?). Requiring the MCU to reset just to get talking with the CAN controller is not a valid fix. 

Adding extra time delay here is not going to help if the CAN controller's voltage supply comes up much later.  The driver has already failed much earlier in the initialisation sequence.  At this point we are waiting 128 MCP2515 clock cycles after the soft reset command.  If you want to make this delay configurable or add 20% more us for a safety margin that would be acceptable.  A major design goal of Zephyr is to fit in small microprocessors so please also take this into consideration.

**Comment by NZSmartie** [comment]
Is there a the possibility of requesting Zephyr to re-initialise a driver from the application? Otherwise I'm not sure what else I can do. 

The voltage rail for the CAN controller becomes stable within 600us of the MCU's voltage rail. however. I'm seeing the CAN controller remain unresponsive for roughly ~15ms, which far exceeds the 128 OSC1 clock cycles. I'm not sure what what's going on exactly. Additionally the system is using a MCP25625 (using has the same SPI protocol as the MCP2515, with a built in CAN transceiver). The documentation regarding 128 OSC1 clock cycles is the same. My guess is there might be an issue with the chip itself or an undocumented start up behaviour.

**Comment by nixward** [comment]
@NZSmartie for your particular system you might be able to use this kconfig option to let the power stabilise before the drivers are initialised: [CONFIG_BOOT_DELAY](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_BOOT_DELAY.html#cmdoption-arg-config-boot-delay)

Note I haven't used this config option myself so I could be wrong.

Similarly I haven't used `k_usleep()` before but you could also use `k_msleep(1)` very safely.

**Comment by alexanderwachter** [comment]
Does it make sense to poll in the init function instead of the config function? Reset and poll mode in a loop with a max timeout.

**Comment by nixward** [comment]
> Does it make sense to poll in the init function instead of the config function? Reset and poll mode in a loop with a max timeout.

I think CONFIG_BOOT_DELAY is better suited to waiting for system voltage stabilisation.  I'm not sure driver code should be addressing this issue.  If every driver has to do this it bulks up the code required per driver.

Either solution will end up with the system boot being delayed.

**Comment by alexanderwachter** [comment]
@nixward it still makes sense to check/pull the device state. That could also help to spot problems with a defect controller. And such a loop is not much overhead.

**Comment by NZSmartie** [comment]
> I think CONFIG_BOOT_DELAY is better suited to waiting for system voltage stabilisation. I'm not sure driver code should be addressing this issue. If every driver has to do this it bulks up the code required per driver.
> 
> Either solution will end up with the system boot being delayed.

The system will be boot delayed in any scenario. Allowing the driver to wait for the device to be ready improves reliability of the driver. Instead of getting a `EIO` and with no way for the application to do anything about it without resetting.

**Comment by nixward** [comment]
@NZSmartie you have raised a valid issue that the driver is not waiting after calling the MCP2515 reset command but we should separate the wait requirements of the driver and the wait requirements of the hardware in your board.  A file called `board.c` can be made within your [custom board definition folder](https://docs.zephyrproject.org/latest/application/index.html#custom-board-devicetree-and-soc-definitions) that's better suited to define the board specific initialisation requirements.

**Comment by nixward** [comment]
See https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/nrf9160dk_nrf52840/board.c as a board.c example.

**Comment by NZSmartie** [comment]
This is a good compromise and I'm happy to go forward with the compromise of board specific initialisation. Give me a moment to test this out and update the PR

File: drivers/can/can_mcp2515.c [hunk 4]
```diff
@@ -271,6 +271,18 @@ const int mcp2515_set_mode(struct device *dev, uint8_t mcp2515_mode)
...
 
+static int mcp2515_get_mode(struct device *dev)
+{
+	uint8_t canstat;
+
+	if (mcp2515_cmd_read_reg(dev, MCP2515_ADDR_CANSTAT, &canstat, 1)) {
+		return -EIO;
+	}
+
+	return (canstat & MCP2515_CANSTAT_MODE_MASK)
```
Comment (karstenkoenig) [inline drivers/can/can_mcp2515.c:282]: Not sure about this, but are we allowed to return both data and return codes in functions at the same time? I believe it'd be better to return 0 on success and the mode in a reference argument. Should be inlined by the compiler anyways with a higher optimization mode.

File: drivers/can/can_mcp2515.c [hunk 5]
```diff
@@ -352,6 +364,18 @@ static int mcp2515_configure(struct device *dev, enum can_mode mode,
...
+	/* Wait 128 OSC1 clock cycles at 1MHz (minimum clock in frequency)
+	* see MCP2515 datasheet section 8.1 Oscillator Start-up Timer
+	*/
+	k_usleep(128);
+
+	ret = mcp2515_get_mode(dev);
+	if (ret != MCP2515_MODE_CONFIGURATION) {
+		LOG_ERR("Device did not reset into configuration mode [%d]",
+			ret);
+		return -EIO;
```
Comment (karstenkoenig) [inline drivers/can/can_mcp2515.c:376]: For cleanliness `goto done` here to clean up the locked mutex. 
Also I just noticed that `done` label sits in the wrong spot, should be further down before the mutex release.  `goto done` is used inconsistently in the rest of this function as well. I will make a small patch to fix the handling.

**Review by karstenkoenig — CHANGES_REQUESTED** [review]

**Comment by karstenkoenig** [comment]
Oh good catch that a 0x00 return is considered a success, good idea to inquire about the correct mode on startup and the delay to make sure it had the spec'ed delay on startup.

On any further delay for startup I agree with @nixward. Adding a delay to each driver is overkill and should be a more generic solution and just waiting in the board startup to account for individual situations might be best :-)

File: drivers/can/can_mcp2515.c [hunk 6]
```diff
@@ -352,6 +371,25 @@ static int mcp2515_configure(struct device *dev, enum can_mode mode,
 		goto done;
 	}
 
+	/* Wait 128 OSC1 clock cycles at 1MHz (minimum clock in frequency)
+	* see MCP2515 datasheet section 8.1 Oscillator Start-up Timer
+	*/
```
Comment (nixward) [inline drivers/can/can_mcp2515.c:376]: Slight nit here, Zephyr follows the Linux kernel commenting style: https://www.kernel.org/doc/html/latest/process/coding-style.html#commenting
The asterisk characters should be aligned.

File: drivers/can/can_mcp2515.c [hunk 7]
```diff
@@ -352,6 +371,25 @@ static int mcp2515_configure(struct device *dev, enum can_mode mode,
...
+	ret = mcp2515_get_mode(dev, &reset_mode);
+	if (ret < 0) {
+		LOG_ERR("Failed to read device mode [%d]",
+			ret);
+		goto done;
+	}
+
+	if (reset_mode != MCP2515_MODE_CONFIGURATION) {
+		LOG_ERR("Device did not reset into configuration mode [%d]",
+			mode);
```
Comment (nixward) [inline drivers/can/can_mcp2515.c:388]: This should be `reset_mode`

File: drivers/can/can_mcp2515.c [hunk 8]
```diff
@@ -352,6 +371,25 @@ static int mcp2515_configure(struct device *dev, enum can_mode mode,
 		goto done;
 	}
 
+	/* Wait 128 OSC1 clock cycles at 1MHz (minimum clock in frequency)
+	 * see MCP2515 datasheet section 8.1 Oscillator Start-up Timer
+	 */
+	k_usleep(128);
```
Comment (nixward) [inline drivers/can/can_mcp2515.c:377]: It's probably worth putting another 128us sleep before the soft reset command.  I don't know how quickly the driver could be initialised after a system is released from reset but if it's quicker than 128us then the soft reset command may be blocked due to the same 128 OSC1 clock cycle requirement.

Comment (nixward) [reply drivers/can/can_mcp2515.c:377]: Thoughts @karstenkoenig ?

Comment (karstenkoenig) [reply drivers/can/can_mcp2515.c:377]: Guess waiting another 128us is safer than not waiting. I won't object nor push for it though, as technically this implies a perfect release from reset between both MCU and MCP2515.

**Review by karstenkoenig — APPROVED** [review]
Thanks!

File: drivers/can/can_mcp2515.c (hunk 8 above)
Comment (nixward) [reply drivers/can/can_mcp2515.c:377]: At the moment there will be somewhere between 0us to >128us between the MCP2515 release from reset or power up.  Adding a 128us sleep before the soft reset will just ensure that 128us is the minimum.  @NZSmartie I wont block the PR either way you choose to go.  Thanks for your fix.

**Review by nixward — APPROVED** [review]

**Review by alexanderwachter — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**